File: frontend/src/entities/troop_training_order.ts
import { Resources, TroopKind, Village } from "."
import * as serverV1Types from "../../lib/protobuf/server/v1/server_pb"

export default class TroopTrainingOrder {
  id: number
  quantity: number
  timeLeft: number
  cost: Resources
  troopKind: TroopKind

  village: Village

  constructor(order: serverV1Types.Troop_TrainingOrder, village: Village) {
    this.id = order.id
    this.quantity = order.quantity
    this.timeLeft = order.timeLeft
    this.cost = new Resources(order.cost!)
    this.troopKind = order.troopKind as TroopKind

    this.village = village
  }
}

File: cmd/http/model/troop_training_order.go
package model

import (
	"context"
	"war-of-faith/cmd/http/db"
	serverv1 "war-of-faith/pkg/protobuf/server/v1"
)

type TroopTrainingOrder struct {
	Id        uint32     `db:"id"`
	Quantity  uint32     `db:"quantity"`
	TimeLeft  uint32     `db:"time_left"`
	TroopKind Troop_Kind `db:"troop_kind"`

	VillageId uint32 `db:"village_id"`
	village   *Village
}

func (o *TroopTrainingOrder) ToProtobuf(ctx context.Context) (*serverv1.Troop_TrainingOrder, error) {
	return &serverv1.Troop_TrainingOrder{
		Id:        o.Id,
		Quantity:  o.Quantity,
		TimeLeft:  o.TimeLeft,
		Cost:      o.TroopKind.CalculateTrainingCost(o.Quantity, 0).ToProtobuf(),
		TroopKind: string(o.TroopKind),
	}, nil
}

func (o *TroopTrainingOrder) Village(ctx context.Context) (Village, error) {
	if o.village == nil {
		village, found, err := GetVillage(ctx, o.VillageId)
		if err != nil {
			return Village{}, err
		}
		if !found {
			return Village{}, db.ErrNotFound
		}
		o.village = &village
	}
	return *o.village, nil
}
